core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+        let mut storage = storage.start_transaction().await?;
+
+        let mut applied_snapshot_status = storage
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await
+            .map_err(|_| Retryable("Unable to get applied snapshots status".to_string()))?;
+
+        if !storage
+            .blocks_dal()
+            .is_genesis_needed()
+            .await
+            .map_err(|_| Retryable("Unable to fetch genesis state".to_string()))?
+            && applied_snapshot_status.is_none()
+        {
+            return Err(Canceled(
+                "This node has already been initialized without a snapshot".to_string(),
+            ));
+        }
+
+        if let Some(applied_snapshot_status) = applied_snapshot_status.as_ref() {
+            if applied_snapshot_status
+                .storage_logs_chunks_ids_to_process
+                .is_empty()
+            {
+                return Err(Canceled(
+                    "This node has already been initialized from a snapshot".to_string(),
+                ));
+            }
+        } else {
+            applied_snapshot_status =
+                Some(SnapshotsApplier::create_fresh_recovery_status(main_node_client).await?)
+        }
+
+        let mut recovery = Self {
+            connection_pool,
+            blob_store,
+            applied_snapshot_status: applied_snapshot_status.unwrap(),
+        };
+
+        recovery.recover_factory_deps(&mut storage).await?;
+
+        storage
+            .snapshot_recovery_dal()
+            .set_applied_snapshot_status(&recovery.applied_snapshot_status)
+            .await
+            .map_err(|err| Fatal(err.into()))?;
+
+        storage.commit().await.map_err(|err| Fatal(err.into()))?;
+
+        recovery.recover_storage_logs().await?;
+
+        Ok(())
+    }
+
+    async fn create_fresh_recovery_status(
+        main_node_client: Box<dyn SnapshotsApplierMainNodeClient>,
+    ) -> Result<SnapshotRecoveryStatus, SnapshotApplierError> {
+        let snapshot_response = main_node_client
+            .fetch_newest_snapshot()
+            .await
+            .map_err(|_| Retryable("Unable to fetch newest snapshot from main node".to_string()))?;
+        if snapshot_response.is_none() {
+            return Err(Canceled("Main node does not have any ready snapshots, skipping initialization from snapshot!".to_string()));
+        }
+        let snapshot = snapshot_response.unwrap();
+        let l1_batch_number = snapshot.l1_batch_number;
+        tracing::info!(
+            "Found snapshot with data up to l1_batch {}, storage_logs are divided into {} chunk(s)",
+            l1_batch_number,
+            snapshot.storage_logs_chunks.len()
+        );
+
+        let miniblock_root_hash = main_node_client
+            .miniblock_hash(snapshot.miniblock_number)
+            .await
+            .map_err(|err| Fatal(err.into()))?
+            .ok_or_else(|| Fatal(Error::msg("Miniblock is missing")))?;
```

Comment:
This feels like a hack. What if we need more data in `SnapshotRecoveryStatus`? (JFYI, preliminarily we'd need `l1_batch_timestamp`, `miniblock_timestamp` and `protocol_version` to support snapshot recovery for the state keeper.) Maybe, it'd make sense to add missing data to `snapshot_response`?

Reply:
I wasn't aware that additional data would be needed here and it's easier to test it now, I also assumed it would be easy to change this to use full miniblock later, but if you say it's certain it will be needed, I will change it now

Reply:
I'm 99% sure that *at least* these fields are needed; I've surveyed the state keeper logic to arrive at this list. Maybe it's not a complete list.